rootfs: switch CPIO compression from LZMA to ZSTD

Knogle opened the pull request:
We move from LZMA to ZSTD for CPIO compression primarily to improve overall build performance and runtime extraction speed. ZSTD provides significantly better multicore scalability during compression and offers much faster decompression, which is beneficial for both developer workflows and boot-time performance.

In our tests, CPIO images compressed with ZSTD are only slightly larger (≈160 MB vs. ≈140 MB with LZMA), but the reduction in build time and the improved parallelism outweigh the small increase in image size. ZSTD has become the de-facto modern replacement for LZMA in high-performance build environments, especially when frequent rebuilds are required.

This change updates all ShredOS CPIO targets to use ZSTD by default.

Knogle committed:
rootfs: switch CPIO compression from LZMA to ZSTD

We move from LZMA to ZSTD for CPIO compression primarily to improve
overall build performance and runtime extraction speed. ZSTD provides
significantly better multicore scalability during compression and
offers much faster decompression, which is beneficial for both
developer workflows and boot-time performance.

In our tests, CPIO images compressed with ZSTD are only slightly
larger (≈160 MB vs. ≈140 MB with LZMA), but the reduction in build
time and the improved parallelism outweigh the small increase in
image size. ZSTD has become the de-facto modern replacement for LZMA
in high-performance build environments, especially when frequent
rebuilds are required.

This change updates all ShredOS CPIO targets to use ZSTD by default.

PartialVolume commented:
I'd like to see some stats on the use of LZMA vs ZSTD in terms of total time to boot, i.e from selecting ShredOS in the GNU grub menu to when the `Waiting for all USB devices to be initialised` message appears.

The concerning thing for me is that increase in size. Any increase in decompression speed may be completely lost due to the increased size and increased I/O when loading the bzImage from the device,especially on a slow connection, like USB 1,2 or PXE, hence why I'd like to see real world numbers on a variety of systems and boot methods before I merge.

I can see the speed benefits for the developer when compressing the image but not if it makes the overall boot process longer for the end user.

Knogle commented:
Ahoy, yep i think we need to do some testing. But it's quite difficult i think to measure those times.
Btw. we can shrink the image by around 100MB compressed size at least, because there is the `kselftest` folder we don't need. But i will set up some kind of benchmark. Would you like a synthetic benchmark, like a script or program unpacking the CPIO, or a real world one, or both?
Especially old and slow CPUs should benefit from moving to ZTSD as CPU load is decreased drastically.

<img width="1064" height="727" alt="Screenshot From 2025-12-12 12-47-18" src="https://github.com/user-attachments/assets/c2c664dc-4672-4773-8954-4ff2839ff500" />
<img width="1064" height="727" alt="Screenshot From 2025-12-12 12-47-03" src="https://github.com/user-attachments/assets/8f1ae291-7466-4b22-b466-856bd04eed32" />
<img width="1064" height="727" alt="Screenshot From 2025-12-12 12-46-58" src="https://github.com/user-attachments/assets/ff099331-33f6-4eb8-8aa0-789be2a1b8ad" />

PartialVolume commented:
I'd prefer real world tests. This Sony Vaio with USB 2 has a grub load speed of ~3MB/s so an extra 20MB would add about 7 seconds.

Losing the kselftest would certainly knock nearly 50% of the boot time.

Is it possible to remove kselftest without breaking anything?

Knogle commented:
I will run some tests :) Yes. Should be possible. Or we could unselect the `linux-tools` package, as those selftests are part of it. And the package should not be necessary for ShredOS. Honestly I would prefer to get rid of the amdgpu driver as well, as this is another 200MB uncompressed. Noveau should be fine for a program like this.

PartialVolume commented:
Sounds good, can you keep those is separate PRs, changes to graphics driver and anything to do with getting rid of kselftests. Thanks.

PartialVolume commented:
Re amdgpu, I thought the Noveau driver for just for Nvidia cards not AMD Radeon?

Knogle commented:
> Re amdgpu, I thought the Noveau driver for just for Nvidia cards not AMD Radeon?

We will just use a generic framebuffer for all types of GPUs, with `nomodeset` set. So there will be no graphic acceleration, that's fine for just displaying the TUI, and there is A LOT less stuff inside the image. I think we can get it back down to 40-60MB again.

PartialVolume commented:
The problem with doing that, is you can't send the system to sleep and expect the graphics to wake up and be in exactly the same state as when the system went to sleep. You just get a blank screen no matter what the graphics card. The DRM and graphics drivers handle bringing back the graphics system to exactly the same state so nwipe is displayed exactly as it was when the system went to sleep.

Sending the system to sleep is necessary for unfreezing drives especially laptops but also PCs  where you don't want to remove the drive or start pulling power connectors.

PartialVolume commented:
Being able to send the system to sleep will be a necessary part of doing a secure erase on a frozen drive when we add that feature in 0.41

Knogle commented:
Ahoy, unfortunately i'm having difficulties on testing this, it's not that easy unfortunately.
ZSTD seems to be around 2s faster on a low-spec system.
We could also try xz, as xz offers the best compression, but at highest cpu load.
I'm currently working on a kernel patch in order to track the actual decompression time using TSC indexes. Another method would be, chainloading kernels using `kexec` to have exact timings. Do you use the img method oder new iso9xxx hybrid images from @desertwitch

Knogle commented:
Just my first step in order to verify that.

That's the code i use for testing, together with our rootfs.cpio from ShredOS output directory.


```c++
// cpio_compress_benchmark.cpp – Benchmarks Buildroot-style CPIO compression algorithms (via CLI tools) with verbose output
// ---------------------------------------------------------------------------------------------------------------
//  • Creates (optional) synthetic “rootfs-like” directory tree, generates an uncompressed rootfs.cpio (newc),
//    then benchmarks common Buildroot CPIO compressors by compressing that same input file.
//  • Measures *compression throughput* in MB/s and reports compressed size + ratio.
//  • Uses external tools (gzip, bzip2, xz/lzma, lz4, zstd, lzop) just like Buildroot host tooling does.
//  • Compile:  g++ -O3 -std=c++17 cpio_compress_benchmark.cpp -o cpio_comp_bench
//  • Run:      ./cpio_comp_bench            (auto-generate sample + benchmark)
//              ./cpio_comp_bench --input rootfs.cpio
//
// Notes:
//  • For fair CPU-only comparison, this writes compressed output to a temp file (not /dev/null) to count bytes.
//  • If a compressor is missing, it’s skipped with a note.
// ---------------------------------------------------------------------------------------------------------------

#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <unistd.h>

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <optional>
#include <random>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace fs = std::filesystem;

static constexpr uint64_t MiB = 1024ull * 1024ull;

static void die(const std::string& msg) {
    throw std::runtime_error(msg + ": " + std::string(std::strerror(errno)));
}

static bool is_executable_in_path(const std::string& prog) {
    // If prog contains '/', check directly.
    if (prog.find('/') != std::string::npos) {
        return ::access(prog.c_str(), X_OK) == 0;
    }
    const char* path = std::getenv("PATH");
    if (!path) return false;

    std::string p(path);
    std::stringstream ss(p);
    std::string dir;
    while (std::getline(ss, dir, ':')) {
        fs::path cand = fs::path(dir) / prog;
        if (::access(cand.c_str(), X_OK) == 0) return true;
    }
    return false;
}

static uint64_t file_size_bytes(const fs::path& p) {
    std::error_code ec;
    auto sz = fs::file_size(p, ec);
    if (ec) die("file_size(" + p.string() + ")");
    return (uint64_t)sz;
}

static std::string human_bytes(uint64_t b) {
    const char* units[] = {"B","KiB","MiB","GiB","TiB"};
    double v = (double)b;
    size_t u = 0;
    while (v >= 1024.0 && u < 4) { v /= 1024.0; ++u; }
    std::ostringstream os;
    os.setf(std::ios::fixed); os.precision(u == 0 ? 0 : 2);
    os << v << " " << units[u];
    return os.str();
}

static double mbps_decimal(uint64_t bytes, double seconds) {
    // MB/s in decimal (1e6), matching many benchmark printouts
    return (double)bytes / 1e6 / seconds;
}

struct Algo {
    std::string name;              // display name
    std::string prog;              // program name for presence check (e.g., "gzip")
    std::vector<std::string> argv; // argv[0].. (we’ll execvp argv[0])
    std::string out_ext;           // suffix for temp output
};

struct Result {
    std::string name;
    double seconds{};
    double mbps{};
    uint64_t in_bytes{};
    uint64_t out_bytes{};
    double ratio{}; // out/in
};

static int spawn_pipeline_one(const std::vector<std::string>& argv,
                              const fs::path& in_file,
                              const fs::path& out_file,
                              bool verbose) {
    if (argv.empty()) throw std::runtime_error("empty argv");

    int in_fd = ::open(in_file.c_str(), O_RDONLY);
    if (in_fd < 0) die("open(in)");

    int out_fd = ::open(out_file.c_str(), O_CREAT | O_TRUNC | O_WRONLY, 0644);
    if (out_fd < 0) die("open(out)");

    pid_t pid = ::fork();
    if (pid < 0) die("fork()");

    if (pid == 0) {
        // child
        if (::dup2(in_fd, STDIN_FILENO) < 0) _exit(127);
        if (::dup2(out_fd, STDOUT_FILENO) < 0) _exit(127);

        ::close(in_fd);
        ::close(out_fd);

        // Build exec argv
        std::vector<char*> cargv;
        cargv.reserve(argv.size() + 1);
        for (const auto& s : argv) cargv.push_back(const_cast<char*>(s.c_str()));
        cargv.push_back(nullptr);

        ::execvp(cargv[0], cargv.data());
        _exit(127);
    }

    // parent
    ::close(in_fd);
    ::close(out_fd);

    int status = 0;
    if (::waitpid(pid, &status, 0) < 0) die("waitpid()");

    if (verbose) {
        if (WIFEXITED(status)) {
            int rc = WEXITSTATUS(status);
            if (rc != 0) std::cerr << "[debug] " << argv[0] << " exited rc=" << rc << "\n";
        } else if (WIFSIGNALED(status)) {
            std::cerr << "[debug] " << argv[0] << " killed by signal " << WTERMSIG(status) << "\n";
        }
    }
    return status;
}

static uint64_t xorshift64(uint64_t& s) {
    // simple deterministic PRNG (fast, good enough for test data)
    s ^= s << 13;
    s ^= s >> 7;
    s ^= s << 17;
    return s;
}

static void write_file_pattern(const fs::path& p, uint64_t size_bytes, uint64_t seed, double compressible_fraction) {
    std::ofstream f(p, std::ios::binary);
    if (!f) throw std::runtime_error("cannot write " + p.string());

    static constexpr size_t BUF = 1u << 20; // 1 MiB
    std::vector<unsigned char> buf(BUF);

    uint64_t state = seed;
    uint64_t written = 0;

    while (written < size_bytes) {
        size_t chunk = (size_t)std::min<uint64_t>(BUF, size_bytes - written);

        // Mix: some sections are highly compressible (0x00 / small repeating), some are random-ish.
        // This roughly mimics a rootfs containing binaries (less compressible) + text/config (more compressible).
        double pick = (double)(xorshift64(state) & 0xFFFF) / 65535.0;
        if (pick < compressible_fraction) {
            unsigned char v = (unsigned char)(xorshift64(state) & 0xFF);
            std::memset(buf.data(), v, chunk);
        } else {
            for (size_t i = 0; i < chunk; ++i) buf[i] = (unsigned char)(xorshift64(state) & 0xFF);
        }

        f.write(reinterpret_cast<const char*>(buf.data()), (std::streamsize)chunk);
        if (!f) throw std::runtime_error("write failed " + p.string());
        written += chunk;
    }
}

static void make_sample_rootfs_tree(const fs::path& dir, uint64_t total_mib, bool verbose) {
    if (verbose) std::cout << "[verbose] Creating sample rootfs tree at: " << dir << "\n";
    fs::create_directories(dir / "bin");
    fs::create_directories(dir / "sbin");
    fs::create_directories(dir / "etc");
    fs::create_directories(dir / "lib");
    fs::create_directories(dir / "usr/bin");
    fs::create_directories(dir / "usr/lib");
    fs::create_directories(dir / "var/log");
    fs::create_directories(dir / "tmp");

    // Create a mix of many small files + a handful of larger ones.
    // Target size is approximate; we’ll stop once we pass total_mib.
    const uint64_t target = total_mib * MiB;
    uint64_t acc = 0;
    uint64_t seed = 0xC0FFEE123456789ull;

    auto mk_small = [&](const fs::path& p, uint64_t bytes, double comp_frac) {
        write_file_pattern(p, bytes, seed, comp_frac);
        seed += 0x9E3779B97F4A7C15ull;
        acc += bytes;
    };

    // Lots of configs (compressible)
    for (int i = 0; i < 400 && acc < target; ++i) {
        mk_small(dir / "etc" / ("conf_" + std::to_string(i) + ".cfg"), 8 * 1024, 0.85);
    }

    // “Binaries” (less compressible)
    for (int i = 0; i < 200 && acc < target; ++i) {
        mk_small(dir / "usr/bin" / ("app_" + std::to_string(i)), 64 * 1024, 0.20);
    }

    // Libraries (mixed)
    for (int i = 0; i < 120 && acc < target; ++i) {
        mk_small(dir / "usr/lib" / ("lib_" + std::to_string(i) + ".so"), 256 * 1024, 0.35);
    }

    // Some larger blobs
    for (int i = 0; i < 8 && acc < target; ++i) {
        mk_small(dir / "lib" / ("blob_" + std::to_string(i) + ".dat"), 16 * MiB, 0.30);
    }

    // A bit of log/text
    for (int i = 0; i < 50 && acc < target; ++i) {
        mk_small(dir / "var/log" / ("log_" + std::to_string(i) + ".txt"), 128 * 1024, 0.90);
    }

    if (verbose) {
        std::cout << "[verbose] Sample rootfs tree created, approx size: " << human_bytes(acc) << "\n";
    }
}

static void build_cpio_newc(const fs::path& root_dir, const fs::path& out_cpio, bool verbose) {
    if (!is_executable_in_path("cpio")) {
        throw std::runtime_error("cpio not found in PATH (needed to generate rootfs.cpio). Install 'cpio' or pass --input.");
    }

    // Use /bin/sh pipeline to reliably run find|cpio and redirect to out file
    // newc is what Buildroot typically uses for rootfs.cpio (unless configured otherwise).
    std::ostringstream cmd;
    cmd << "set -euo pipefail; "
        << "cd " << std::quoted(root_dir.string()) << " ; "
        << "find . -print0 | cpio --null -o -H newc --quiet > " << std::quoted(out_cpio.string());

    if (verbose) std::cout << "[verbose] Generating uncompressed CPIO (newc): " << out_cpio << "\n";
    int rc = std::system(cmd.str().c_str());
    if (rc != 0) throw std::runtime_error("cpio generation failed (rc=" + std::to_string(rc) + ")");
}

static Result bench_one(const Algo& a, const fs::path& in_cpio, const fs::path& out_dir, bool verbose) {
    Result r;
    r.name = a.name;
    r.in_bytes = file_size_bytes(in_cpio);

    fs::path out_file = out_dir / ("out" + a.out_ext);

    if (verbose) {
        std::cout << "[verbose] Benchmarking: " << a.name << "\n";
        std::cout << "         exec: ";
        for (const auto& s : a.argv) std::cout << s << " ";
        std::cout << "\n";
    }

    auto t0 = std::chrono::steady_clock::now();
    int status = spawn_pipeline_one(a.argv, in_cpio, out_file, verbose);
    auto t1 = std::chrono::steady_clock::now();

    if (!(WIFEXITED(status) && WEXITSTATUS(status) == 0)) {
        throw std::runtime_error(a.name + " failed (non-zero exit)");
    }

    r.seconds = std::chrono::duration<double>(t1 - t0).count();
    r.out_bytes = file_size_bytes(out_file);
    r.ratio = (r.in_bytes == 0) ? 0.0 : (double)r.out_bytes / (double)r.in_bytes;
    r.mbps = mbps_decimal(r.in_bytes, r.seconds);

    // Keep output for size measurement, but you can delete it if you want:
    // fs::remove(out_file);

    if (verbose) {
        std::cout << "[verbose] Done: " << a.name << " in " << r.seconds << " s"
                  << " | " << r.mbps << " MB/s"
                  << " | out=" << human_bytes(r.out_bytes)
                  << " | ratio=" << r.ratio << "\n";
    }
    return r;
}

static void print_usage(const char* argv0) {
    std::cout <<
        "Usage:\n"
        "  " << argv0 << " [--input rootfs.cpio] [--sample-mib N] [--workdir DIR] [--quiet]\n"
        "\n"
        "Defaults:\n"
        "  --sample-mib 512     (if --input is not provided)\n"
        "  --workdir   /tmp/... (auto temp dir)\n";
}

int main(int argc, char** argv) {
    try {
        std::optional<fs::path> input_cpio;
        uint64_t sample_mib = 512;
        std::optional<fs::path> workdir;
        bool verbose = true;

        for (int i = 1; i < argc; ++i) {
            std::string a = argv[i];
            if (a == "--input" && i + 1 < argc) {
                input_cpio = fs::path(argv[++i]);
            } else if (a == "--sample-mib" && i + 1 < argc) {
                sample_mib = std::stoull(argv[++i]);
            } else if (a == "--workdir" && i + 1 < argc) {
                workdir = fs::path(argv[++i]);
            } else if (a == "--quiet") {
                verbose = false;
            } else if (a == "--help" || a == "-h") {
                print_usage(argv[0]);
                return 0;
            } else {
                throw std::runtime_error("Unknown arg: " + a);
            }
        }

        fs::path wd;
        if (workdir) {
            wd = *workdir;
            fs::create_directories(wd);
        } else {
            // create temp dir
            std::string templ = "/tmp/cpio-comp-bench.XXXXXX";
            std::vector<char> t(templ.begin(), templ.end());
            t.push_back('\0');
            char* p = ::mkdtemp(t.data());
            if (!p) die("mkdtemp()");
            wd = fs::path(p);
        }

        if (verbose) {
            std::cout << "CPIO Compression Benchmark (Buildroot-style)\n";
            std::cout << "[verbose] Workdir: " << wd << "\n";
        }

        fs::path in_cpio;
        if (input_cpio) {
            in_cpio = *input_cpio;
            if (!fs::exists(in_cpio)) throw std::runtime_error("Input not found: " + in_cpio.string());
            if (verbose) std::cout << "[verbose] Using input CPIO: " << in_cpio << " (" << human_bytes(file_size_bytes(in_cpio)) << ")\n";
        } else {
            fs::path root = wd / "rootfs";
            fs::create_directories(root);
            make_sample_rootfs_tree(root, sample_mib, verbose);

            in_cpio = wd / "rootfs.cpio";
            build_cpio_newc(root, in_cpio, verbose);

            if (verbose) std::cout << "[verbose] rootfs.cpio size: " << human_bytes(file_size_bytes(in_cpio)) << "\n";
        }

        // Define algorithms roughly matching Buildroot host compressors.
        // All use stdout (-c) so we can redirect.
        std::vector<Algo> algos = {
            // "none" baseline (still measures I/O + copy through)
            {"none (cat)", "cat", {"cat"}, ".none"},

            // gzip levels (Buildroot commonly uses gzip for initramfs/rootfs, depending on config)
            {"gzip -1", "gzip", {"gzip", "-c", "-1"}, ".gz"},
            {"gzip -6", "gzip", {"gzip", "-c", "-6"}, ".gz"},
            {"gzip -9", "gzip", {"gzip", "-c", "-9"}, ".gz"},

            // bzip2
            {"bzip2", "bzip2", {"bzip2", "-c"}, ".bz2"},

            // xz (LZMA2) and "lzma" container (for legacy LZMA-in-initramfs style)
            {"xz -0", "xz", {"xz", "-c", "-0"}, ".xz"},
            {"xz -6", "xz", {"xz", "-c", "-6"}, ".xz"},
            {"xz -9e", "xz", {"xz", "-c", "-9e"}, ".xz"},
            {"lzma (xz --format=lzma -6)", "xz", {"xz", "--format=lzma", "-c", "-6"}, ".lzma"},

            // lz4
            {"lz4", "lz4", {"lz4", "-c", "-q"}, ".lz4"},

            // lzo via lzop (common userland interface)
            {"lzop", "lzop", {"lzop", "-c", "-q"}, ".lzo"},

            // zstd
            {"zstd -1", "zstd", {"zstd", "-c", "-q", "-1"}, ".zst"},
            {"zstd -3", "zstd", {"zstd", "-c", "-q", "-3"}, ".zst"},
            {"zstd -19", "zstd", {"zstd", "-c", "-q", "-19"}, ".zst"},
        };

        // Filter by availability
        std::vector<Algo> avail;
        for (const auto& a : algos) {
            if (is_executable_in_path(a.prog)) {
                avail.push_back(a);
            } else if (verbose) {
                std::cout << "[verbose] Skipping (not found): " << a.name << " (needs '" << a.prog << "')\n";
            }
        }
        if (avail.empty()) throw std::runtime_error("No compressors found in PATH.");

        if (verbose) {
            std::cout << "[verbose] Starting benchmark over " << avail.size() << " algorithm variants...\n";
            std::cout << "         Input bytes: " << human_bytes(file_size_bytes(in_cpio)) << "\n";
        }

        fs::path out_dir = wd / "out";
        fs::create_directories(out_dir);

        std::vector<Result> results;
        results.reserve(avail.size());

        for (size_t i = 0; i < avail.size(); ++i) {
            const char sp[] = {'|','/','-','\\'};
            if (verbose) std::cout << "\n[" << sp[i % 4] << "] " << avail[i].name << "\n";

            results.push_back(bench_one(avail[i], in_cpio, out_dir, verbose));
        }

        // Sort by throughput (descending)
        std::sort(results.begin(), results.end(), [](const Result& a, const Result& b) {
            return a.mbps > b.mbps;
        });

        std::cout << "\n=== Results (sorted by throughput) ===\n";
        std::cout << "Input: " << human_bytes(results.front().in_bytes) << "\n\n";

        std::cout << "  Rank | Algorithm                     | Time (s) |  MB/s   | Out Size    | Ratio\n";
        std::cout << "  -----+-------------------------------+----------+---------+-------------+--------\n";
        for (size_t i = 0; i < results.size(); ++i) {
            const auto& r = results[i];
            std::ostringstream name;
            name << r.name;
            std::string n = name.str();
            if (n.size() > 29) n = n.substr(0, 29);

            std::cout << "  " << (i + 1);
            if (i + 1 < 10) std::cout << "    | ";
            else std::cout << "   | ";

            std::cout << n;
            if (n.size() < 29) std::cout << std::string(29 - n.size(), ' ');
            std::cout << " | ";

            std::cout.setf(std::ios::fixed);
            std::cout.precision(3);
            std::cout << r.seconds << "  | ";

            std::cout.precision(1);
            std::cout << r.mbps << " | ";

            std::cout << human_bytes(r.out_bytes);
            std::string hs = human_bytes(r.out_bytes);
            if (hs.size() < 11) std::cout << std::string(11 - hs.size(), ' ');
            std::cout << " | ";

            std::cout.precision(3);
            std::cout << r.ratio << "\n";
        }

        std::cout << "\nFastest (this run): " << results.front().name
                  << " @ " << results.front().mbps << " MB/s\n";

        if (verbose) {
            std::cout << "\n[verbose] Outputs are in: " << (wd / "out") << "\n";
            std::cout << "[verbose] Tip: rerun with --input <your real rootfs.cpio> for most realistic comparison.\n";
        }

        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Fatal: " << e.what() << "\n";
        return 1;
    }
}

```

Compile and run:

```
g++ -O3 -std=c++17 cpio_compress_benchmark.cpp -o cpio_comp_bench
./cpio_comp_bench --sample-mib 1024
# with actual rootfs
./cpio_comp_bench --input /path/to/rootfs.cpio
```

Those results for our `rootfs.cpio` show the compression rate and speed for compression and decompression.
```

CPIO Compression Benchmark (Buildroot-style)
[verbose] Workdir: "/tmp/cpio-comp-bench.v3rrc2"
[verbose] Using input CPIO: "rootfs.cpio" (776.80 MiB)
[verbose] Starting benchmark over 14 algorithm variants...
         Input bytes: 776.80 MiB

[|] none (cat)
[verbose] Benchmarking: none (cat)
         exec: cat 
[verbose] Done: none (cat) in 0.0199862 s | 40754.9 MB/s | out=776.80 MiB | ratio=1

[/] gzip -1
[verbose] Benchmarking: gzip -1
         exec: gzip -c -1 
[verbose] Done: gzip -1 in 7.13029 s | 114.236 MB/s | out=266.67 MiB | ratio=0.343288

[-] gzip -6
[verbose] Benchmarking: gzip -6
         exec: gzip -c -6 
[verbose] Done: gzip -6 in 19.4349 s | 41.911 MB/s | out=244.89 MiB | ratio=0.315259

[\] gzip -9
[verbose] Benchmarking: gzip -9
         exec: gzip -c -9 
[verbose] Done: gzip -9 in 60.9222 s | 13.3701 MB/s | out=243.78 MiB | ratio=0.31383

[|] bzip2
[verbose] Benchmarking: bzip2
         exec: bzip2 -c 
[verbose] Done: bzip2 in 38.0675 s | 21.3972 MB/s | out=218.74 MiB | ratio=0.281591

[/] xz -0
[verbose] Benchmarking: xz -0
         exec: xz -c -0 
[verbose] Done: xz -0 in 1.46016 s | 557.839 MB/s | out=206.52 MiB | ratio=0.265865

[-] xz -6
[verbose] Benchmarking: xz -6
         exec: xz -c -6 
[verbose] Done: xz -6 in 17.921 s | 45.4515 MB/s | out=150.19 MiB | ratio=0.193338

[\] xz -9e
[verbose] Benchmarking: xz -9e
         exec: xz -c -9e 
[verbose] Done: xz -9e in 63.4474 s | 12.838 MB/s | out=138.73 MiB | ratio=0.178588

[|] lzma (xz --format=lzma -6)
[verbose] Benchmarking: lzma (xz --format=lzma -6)
         exec: xz --format=lzma -c -6 
[verbose] Done: lzma (xz --format=lzma -6) in 122.372 s | 6.65623 MB/s | out=146.93 MiB | ratio=0.189153

[/] lz4
[verbose] Benchmarking: lz4
         exec: lz4 -c -q 
[verbose] Done: lz4 in 0.369997 s | 2201.47 MB/s | out=337.18 MiB | ratio=0.434066

[-] lzop
[verbose] Benchmarking: lzop
         exec: lzop -c -q 
[verbose] Done: lzop in 1.24585 s | 653.798 MB/s | out=336.09 MiB | ratio=0.432659

[\] zstd -1
[verbose] Benchmarking: zstd -1
         exec: zstd -c -q -1 
[verbose] Done: zstd -1 in 0.387471 s | 2102.18 MB/s | out=254.20 MiB | ratio=0.327235

[|] zstd -3
[verbose] Benchmarking: zstd -3
         exec: zstd -c -q -3 
[verbose] Done: zstd -3 in 0.794594 s | 1025.1 MB/s | out=209.19 MiB | ratio=0.269291

[/] zstd -19
[verbose] Benchmarking: zstd -19
         exec: zstd -c -q -19 
[verbose] Done: zstd -19 in 42.1116 s | 19.3423 MB/s | out=159.67 MiB | ratio=0.205547

=== Results (sorted by throughput) ===
Input: 776.80 MiB

  Rank | Algorithm                     | Time (s) |  MB/s   | Out Size    | Ratio
  -----+-------------------------------+----------+---------+-------------+--------
  1    | none (cat)                    | 0.020  | 40754.9 | 776.80 MiB  | 1.000
  2    | lz4                           | 0.370  | 2201.5 | 337.18 MiB  | 0.434
  3    | zstd -1                       | 0.387  | 2102.2 | 254.20 MiB  | 0.327
  4    | zstd -3                       | 0.795  | 1025.1 | 209.19 MiB  | 0.269
  5    | lzop                          | 1.246  | 653.8 | 336.09 MiB  | 0.433
  6    | xz -0                         | 1.460  | 557.8 | 206.52 MiB  | 0.266
  7    | gzip -1                       | 7.130  | 114.2 | 266.67 MiB  | 0.343
  8    | xz -6                         | 17.921  | 45.5 | 150.19 MiB  | 0.193
  9    | gzip -6                       | 19.435  | 41.9 | 244.89 MiB  | 0.315
  10   | bzip2                         | 38.067  | 21.4 | 218.74 MiB  | 0.282
  11   | zstd -19                      | 42.112  | 19.3 | 159.67 MiB  | 0.206
  12   | gzip -9                       | 60.922  | 13.4 | 243.78 MiB  | 0.314
  13   | xz -9e                        | 63.447  | 12.8 | 138.73 MiB  | 0.179
  14   | lzma (xz --format=lzma -6)    | 122.372  | 6.7 | 146.93 MiB  | 0.189
```
